src/psyclone/psyGen.py

```diff
@@ -2771,19 +2740,65 @@ def gen_code(self, parent):
 
         :param parent: The parent of this kernel call in the f2pygen AST.
         :type parent: :py:calls:`psyclone.f2pygen.LoopGen`
+
+        :raises NotImplementedError: if there is a name clash preventing to \
+            module-inline the kernel without changing its name.
         '''
-        from psyclone.f2pygen import CallGen, UseGen
+        from psyclone.f2pygen import CallGen, UseGen, PSyIRGen
 
         # If the kernel has been transformed then we rename it. If it
         # is *not* being module inlined then we also write it to file.
         self.rename_and_write()
 
-        parent.add(CallGen(parent, self._name,
-                           self.arguments.raw_arg_list()))
+        # Add the subroutine call with the necessary arguments
+        arguments = self.arguments.raw_arg_list()
+        parent.add(CallGen(parent, self._name, arguments))
 
+        # Also add the subroutine declaration, this can just be the import
+        # statement, or the whole subroutine inlined into the module.
         if not self.module_inline:
             parent.add(UseGen(parent, name=self._module_name, only=True,
                               funcnames=[self._name]))
+        else:
+            # Find the root f2pygen module
+            module = parent
+            while module.parent:
+                module = module.parent
+
+            # Check for name clashes
+            try:
+                existing_symbol = self.scope.symbol_table.lookup(self._name)
+            except KeyError:
+                existing_symbol = None
+
+            if not existing_symbol:
+                # If it doesn't exist already, module-inline the subroutine
+                self.root.symbol_table.add(RoutineSymbol(self._name))
+                module.add(PSyIRGen(module, self.get_kernel_schedule()))
+            else:
+                # If the symbol already exists, make sure it refers
+                # to the exact same subroutine.
+                if not isinstance(existing_symbol, RoutineSymbol):
+                    raise NotImplementedError(
+                        "Can not module-inline subroutine '{0}' because symbol"
+                        "'{1}' with the same name already exists and changing"
+                        " names of module-inlined subroutines is not "
+                        "implemented yet.".format(self._name, existing_symbol))
+
+                # Make sure the generated code is an exact match
+                found = False
+                search = PSyIRGen(module, self.get_kernel_schedule()).root
```

Comment:
I have to admit that I don't understand this code. Can you not just do `module.root` rather than creating a `PSyIRGen` only to then look at its root? Please add comments - what does checking that the two roots are the same mean?
Also, as an implementation thing you can do away with the `found` variable:

    for child in module.children:
        if blah:
            break
    else:
         # failed to find blah

Reply:
Again I think the difficulty here could be on interpreting `root` as it doesn't return an ancestor but the fparser equivalent to the associated f2pygen node.

What I am doing is create the f2pygen  (which in turn creates the fparser1 version) of the kernel_schedule and then compare it to any of the fparser1 tree associated to f2pygen nodes under the module. If found it means there is already a subroutine with he exact same name and implementation (fparser1 tree), so we can safely continue without inserting it again.

I added a comment to try to explain all this.

Also the `for else` is a good suggestion that I always forget.

Reply:
Aha. I get it now. Thanks for adding comments.